src/librbd/operation/SparsifyRequest.cc

```diff
@@ -19,6 +20,58 @@
 namespace librbd {
 namespace operation {
 
+namespace {
+
+bool may_be_trimmed(const std::map<uint64_t,uint64_t> &extent_map,
+                    const bufferlist &bl, size_t sparse_size,
+                    uint64_t *new_end_ptr) {
+  if (extent_map.empty()) {
+    *new_end_ptr = 0;
+    return true;
+  }
+
+  uint64_t end = extent_map.rbegin()->first + extent_map.rbegin()->second;
+  uint64_t new_end = end;
+  uint64_t bl_off = bl.length();
+
+  for (auto it = extent_map.rbegin(); it != extent_map.rend(); it++) {
+    auto off = it->first;
+    auto len = it->second;
+
+    new_end = p2roundup<uint64_t>(off + len, sparse_size);
+
+    uint64_t extent_left = len;
+    uint64_t sub_len = len % sparse_size;
+    if (sub_len == 0) {
+      sub_len = sparse_size;
+    }
+    while (extent_left > 0) {
+      ceph_assert(bl_off >= sub_len);
+      bl_off -= sub_len;
+      bufferlist sub_bl;
+      sub_bl.substr_of(bl, bl_off, sub_len);
+      if (!sub_bl.is_zero()) {
+        break;
+      }
+      new_end -= sparse_size;
```

Comment:
If you have a ```sub_len < sparse_size```, couldn't this result in incorrectly truncating off up to ```sparse_size - sub_len``` amount of non-zeroed data in the worst case? I think you would always need to start at a ```sparse_size```-rounded end position and keep substracting ```sparse_size``` amounts. For all extents within the the new range, verify that all matching ```sub_bl```s are zeroed (might be more than one).

Reply:
I thought I already do this (start at a sparse_size-rounded end position), see the line 41? Alternatively, I could not do that roundup there and subtract `sub_len` from `new_end` and then do just one `new_end` roundup at the end (before comparing with  `end`).

And I think I am verifying all matching `sub_bl`'s are zeroed, staring from the last one down, until a non-zeroed `sub_bl` is meat.

Reply:
Indeed -- I guess I missed that part.